gateway/src/service.rs

```diff
@@ -0,0 +1,351 @@
+}
+
+const DB_PATH: &'static str = "gateway.sqlite";
+
+use crate::API_PORT;
+use crate::{auth::Key, AccountName};
+use crate::auth::User;
+use crate::client::Client;
+
+impl GatewayService {
+    pub async fn init() -> Arc<Self> {
+        let docker = Docker::connect_with_http_defaults().unwrap();
+
+        let hyper = HyperClient::new();
+        if !StdPath::new(DB_PATH).exists() {
+            Sqlite::create_database(DB_PATH).await.unwrap();
+        }
+
+        let db = SqlitePool::connect(DB_PATH).await.unwrap();
+
+        query("CREATE TABLE IF NOT EXISTS projects (project_name TEXT PRIMARY KEY, account_name TEXT NOT NULL, project_state JSON NOT NULL)")
+            .execute(&db)
+            .await
+            .unwrap();
+
+        query("CREATE TABLE IF NOT EXISTS accounts (account_name TEXT PRIMARY KEY, key TEXT UNIQUE, super_user BOOLEAN DEFAULT FALSE)")
+            .execute(&db)
+            .await
+            .unwrap();
+
+        let (work, mut queue) = channel(256);
+
+        let service = Arc::new(Self {
+            docker,
+            hyper,
+            db,
+            work: work.clone()
+        });
+
+        let worker = Arc::clone(&service);
+
+        tokio::spawn(async move {
+            while let Some(Work {
+                project_name,
+                project,
+                account_name
+            }) = queue.recv().await
+            {
+                let current_state = project.state();
+                debug!(
+                    "picking up project_name={} in state={}",
+                    project_name, current_state
+                );
+                // TODO panicking loses the worker thread and bricks the service
+                let next = project.next(&worker.context()).await.unwrap();
+                worker.update_project(&project_name, &next).await.unwrap();
+                debug!(
+                    "left project_name={} in new state={} for account={}",
+                    project_name,
+                    next.state(),
+                    account_name
+                );
+                if current_state != next.state() {
+                    debug!(
+                        "queuing more work for project_name={} (not done)",
+                        project_name
+                    );
+                    // TODO may deadlock?
+                    work.send(Work {
+                        project_name,
+                        project: next,
+                        account_name
+                    })
+                    .await
+                    .ok()
+                    .expect("failed to queue work");
+                } else {
+                    debug!("worker dropping project_name={} (done)", project_name);
+                }
+            }
+        });
+
+        // Queue up all the projects for reconciliation
+        for Work {
+            project_name,
+            project,
+            account_name
+        } in service.iter_projects().await
```

Comment:
It seems we will also start up projects users have deleted since we never remove them from the database and since `iter_projects()` will return projects in _all_ states.